README.md

```diff
@@ -1,214 +1,324 @@
-
-**Note:** Since pushing comes right at the end of the CI, in many cases you don't need to have a valid registry to test the CI on a fork.  It will fail on the push step, but all other steps will clearly work and you can know it should safely merge back into the main repo.
+
+1. fork the 'kubeflow-containers' repo
+2. Modify the CI with
+
+- REGISTRY: (your own dockerhub repo, eg: "j-smith" (no need for the full url))
+- Change
+  ```
+  - uses: azure/docker-login@v1
+    with:
+      login-server: ${{ env.REGISTRY_NAME }}.azurecr.io
+      username: ${{ secrets.REGISTRY_USERNAME }}
+      password: ${{ secrets.REGISTRY_PASSWORD }}
+  ```
+  to
+  ```
+  - uses: docker/login-action@v1
+    with:
+      username: ${{ secrets.REGISTRY_USERNAME }}
+      password: ${{ secrets.REGISTRY_PASSWORD }}
+  ```
+
+3. In your forked repo, define secrets for REGISTRY_USERNAME and REGISTRY_PASSWORD with your dockerhub credentials (you should use an API token, not your actual dockerhub password)
+
+---
 
 ## Other Development Notes
 
-### The `latest` and `v1` tags for the master branch
+### Github CI
+
+The Github workflow is set up to build the images and their dependant stages.
+See below for a flowchart of this build.
+
+The main workflow is `docker.yaml`,
+it controls the stage build order, and what triggers the CI.
+(Pushes to master, pushes to an open pull-request, and nightly builds)
+
+The building of a stage is controled by `docker-steps.yaml`.
+It checks if there are changes to the stage or dependant stages.
+Builds a new image if there are changes, 
+or pulls a copy of the existing image if not.
+Testing will be performed if this is the final stage in the build of an image.
 
-These are intended to be `long-lived` in that they will not change. Subsequent pushes will clobber the previous `jupyterlab-cpu:latest` image. Previously when we built and pushed to master with updates to an image, we would need to go and change the spawner to use that new image. This will allow us to have them reference `jupyterlab-cpu:latest` and remove us from needing to update it. Additionally, upon changing the `ImagePullPolicy` to `Always` we could do restarts of workloads and then guarantee that users are on the 'latest' image.
+![A flowchart of the Github CI workflow](./docs/images/Workflows.png)
 
-The `v1` tag is intended for when we encounter a breaking change but still want to support the features of that current image. We would then branch off and modify the CI as well as increment the tag. 
+### The `v2` and `latest` tags for the master branch
+
+
+These tags are intended to be `long-lived` in that they will not change.
+Subsequent pushes will clobber the previous `IMAGENAME:v2` image.
+This means that `IMAGENAME:v2` will be updated automatically as changes are made,
+so updates to the tag are not needed.
+
+A new `v3` tag will be created for adding these breaking changes.
```

Comment:
v3 tag sounds new. Is this tag actually used by anything? Or is it a tag that is set in the workflows?

Reply:
It is new as a mention, but it's not used at all.   
This section describes the future tag when breaking changes are encountered.

It's a general update to the docs. 
AAW used to say that it was `v1` and `v2` would be used when breaking changes where encountered. But the Zone existed and was already using `v2`. So a note was made, and if breaking changes are made we will move to some `v3` tag so that the existing images can still be supported on `v1` or `v2`

Reply:
ah gotcha. Didnt know aaw was talking about v2.